File: android/app/src/main/java/com/frenchnclc7/app/data/Plan.kt
package com.frenchnclc7.app.data

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.builtins.ListSerializer

const val TOTAL_DAYS = 301

/** The five study sections. `storageKey` is where the web app saves that section's progress. */
enum class PlanSection(val id: String, val title: String, val tagline: String, val storageKey: String, val file: String) {
    ANKI("anki", "Anki vocabulary", "Daily flashcards, both directions", "progress", "anki"),
    GRAMMAR("grammar", "Grammar book", "Grammaire Progressive du Français", "grammar-progress", "grammar"),
    KWIZIQ("kwiziq", "Kwiziq", "Daily lesson links", "kwiziq-progress", "kwiziq"),
    TV5("tv5monde", "TV5MONDE", "Daily listening links", "tv5-progress", "tv5"),
    WRITING("writing", "Writing", "Daily writing task", "writing-progress", "writing"),
}

@Serializable
data class AnkiCard(val i: String, val f: String, val e: String)

@Serializable
data class AnkiDay(val d: Int, val w: Int, val c: List<AnkiCard>)

/** Grammar, Kwiziq, TV5 and Writing all share this shape (`l` is TV5's level badge). */
@Serializable
data class TextDay(val d: Int, val w: Int, val x: String, val l: String? = null, val e: List<GrammarEntryRaw> = emptyList())

/** One book reference on a Grammar day: which book, which chapters, and the chapter titles. */
@Serializable
data class GrammarEntryRaw(val b: String, val c: List<Int> = emptyList(), val t: String? = null)

/** The chapters to open in one grammar book for a day (a day can name several books). */
data class BookChapters(val book: String, val chapters: List<Int>) {
    /** "A1-A2 ch. 1, 4" */
    val label: String get() = "$book ch. " + chapters.joinToString(", ")
}

object GrammarBooks {
    /** Groups a day's entries by book, keeping the order books first appear and every chapter in order (same as the web app). */
    fun group(entries: List<GrammarEntryRaw>): List<BookChapters> {
        val order = LinkedHashMap<String, MutableList<Int>>()
        for (e in entries) order.getOrPut(e.b) { mutableListOf() }.addAll(e.c)
        return order.map { (book, chapters) -> BookChapters(book, chapters) }
    }
}

/** What one day of one section shows. `chapters` is only used by Grammar. */
data class DayContent(
    val day: Int,
    val week: Int,
    val badge: String?,
    val text: String,
    val cards: List<AnkiCard>,
    val chapters: List<BookChapters> = emptyList(),
)

private val json = Json { ignoreUnknownKeys = true }

/** Parses the exported plan files (see tools/export-plan-data.mjs). Pure, so it is unit-tested. */
object PlanParser {
    fun anki(text: String): List<DayContent> =
        json.decodeFromString(ListSerializer(AnkiDay.serializer()), text)
            .map { DayContent(it.d, it.w, null, it.c.size.toString() + " new cards", it.c) }

    fun textDays(text: String): List<DayContent> =
        json.decodeFromString(ListSerializer(TextDay.serializer()), text)
            .map { DayContent(it.d, it.w, it.l, it.x, emptyList(), GrammarBooks.group(it.e)) }
}

/** Loads the plan JSON bundled in the app's assets, once per section. */
class PlanRepository(private val context: Context) {
    private val cache = mutableMapOf<PlanSection, List<DayContent>>()

    fun days(section: PlanSection): List<DayContent> = cache.getOrPut(section) {
        val text = context.assets.open("plan/" + section.file + ".json").bufferedReader(Charsets.UTF_8).use { it.readText() }
        if (section == PlanSection.ANKI) PlanParser.anki(text) else PlanParser.textDays(text)
    }
}

File: ios/FrenchNCLC7/Models/Plan.swift
import Foundation

let totalDays = 301

/// The five study sections. `storageKey` is where the web app saves that section's progress.
enum PlanSection: String, CaseIterable, Identifiable {
    case anki, grammar, kwiziq, tv5, writing

    var id: String { rawValue }

    var title: String {
        switch self {
        case .anki: return "Anki vocabulary"
        case .grammar: return "Grammar book"
        case .kwiziq: return "Kwiziq"
        case .tv5: return "TV5MONDE"
        case .writing: return "Writing"
        }
    }

    var tagline: String {
        switch self {
        case .anki: return "Daily flashcards, both directions"
        case .grammar: return "Grammaire Progressive du Français"
        case .kwiziq: return "Daily lesson links"
        case .tv5: return "Daily listening links"
        case .writing: return "Daily writing task"
        }
    }

    var storageKey: String {
        switch self {
        case .anki: return "progress"
        case .grammar: return "grammar-progress"
        case .kwiziq: return "kwiziq-progress"
        case .tv5: return "tv5-progress"
        case .writing: return "writing-progress"
        }
    }

    /// Name of the bundled plan file (see tools/export-plan-data.mjs).
    var file: String { rawValue }
}

struct AnkiCard: Codable, Hashable {
    let i: String
    let f: String
    let e: String
}

private struct AnkiDayRaw: Codable {
    let d: Int
    let w: Int
    let c: [AnkiCard]
}

/// Grammar, Kwiziq, TV5 and Writing all share this shape (`l` is TV5's level badge).
private struct TextDayRaw: Codable {
    let d: Int
    let w: Int
    let x: String
    let l: String?
}

/// What one day of one section shows.
struct DayContent: Identifiable, Equatable {
    let day: Int
    let week: Int
    let badge: String?
    let text: String
    let cards: [AnkiCard]
    var id: Int { day }
}

/// Parses the exported plan files. Pure, so it is unit-tested.
enum PlanParser {
    static func anki(_ data: Data) throws -> [DayContent] {
        try JSONDecoder().decode([AnkiDayRaw].self, from: data).map {
            DayContent(day: $0.d, week: $0.w, badge: nil, text: "\($0.c.count) new cards", cards: $0.c)
        }
    }

    static func textDays(_ data: Data) throws -> [DayContent] {
        try JSONDecoder().decode([TextDayRaw].self, from: data).map {
            DayContent(day: $0.d, week: $0.w, badge: $0.l, text: $0.x, cards: [])
        }
    }
}

/// Loads the plan JSON bundled in the app, once per section.
final class PlanRepository {
    private var cache: [PlanSection: [DayContent]] = [:]
    private let bundle: Bundle

    init(bundle: Bundle = .main) {
        self.bundle = bundle
    }

    func days(_ section: PlanSection) -> [DayContent] {
        if let cached = cache[section] { return cached }
        guard let url = bundle.url(forResource: section.file, withExtension: "json"),
              let data = try? Data(contentsOf: url) else { return [] }
        let parsed = (section == .anki ? try? PlanParser.anki(data) : try? PlanParser.textDays(data)) ?? []
        cache[section] = parsed
        return parsed
    }
}
